crates/pool-manager/src/consensus.rs

```diff
@@ -0,0 +1,460 @@
+                let _ = tx.send(res);
+            }
+        }
+    }
+
+    fn on_pool_events(&mut self, orders: Vec<PoolInnerEvent>, waker: impl Fn() -> Waker) {
+        use angstrom_types::network::ReputationChangeKind;
+
+        let valid_orders = orders
+            .into_iter()
+            .filter_map(|order| match order {
+                PoolInnerEvent::Propagation(order) => Some(order),
+                PoolInnerEvent::BadOrderMessages(o) => {
+                    o.into_iter().for_each(|peer| {
+                        self.network
+                            .peer_reputation_change(peer, ReputationChangeKind::InvalidOrder);
+                    });
+                    None
+                }
+                PoolInnerEvent::HasTransitionedToNewBlock(block) => {
+                    self.global_sync
+                        .sign_off_on_block(MODULE_NAME, block, Some(waker()));
+                    None
+                }
+                PoolInnerEvent::None => None
+            })
+            .collect::<Vec<_>>();
+
+        self.broadcast_orders_to_peers(valid_orders);
+    }
+}
+
+impl<V, GlobalSync, NH> Future for ConsensusPoolManager<V, GlobalSync, NH>
+where
+    V: OrderValidatorHandle<Order = AllOrders> + Unpin,
+    GlobalSync: BlockSyncConsumer,
+    NH: NetworkHandle<Events<'static> = UnboundedReceiverStream<StromNetworkEvent>>
+        + Unpin
+        + 'static
+{
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.get_mut();
+
+        // High priority: pull all eth events
+        while let Poll::Ready(Some(eth)) = this.eth_network_events.poll_next_unpin(cx) {
+            this.on_eth_event(eth, cx.waker().clone());
+        }
+
+        // High priority: poll underlying pool. This is the validation process that's
+        // being polled
+        while let Poll::Ready(Some(orders)) = this.order_indexer.poll_next_unpin(cx) {
+            this.on_pool_events(orders, || cx.waker().clone());
+        }
+
+        // Medium priority: Poll network/peer related events - consensus mode specific
+        while let Poll::Ready(Some(event)) = this.strom_network_events.poll_next_unpin(cx) {
+            this.on_network_event(event);
+        }
+
+        // Medium priority: Poll incoming network order events - consensus mode specific
+        if this.global_sync.can_operate() {
+            while let Poll::Ready(Some(event)) = this.order_events.poll_next_unpin(cx) {
+                this.on_network_order_event(event);
+            }
+        }
+
+        // Low priority: halt dealing with these till we have synced
+        if this.global_sync.can_operate() {
+            // drain commands
+            while let Poll::Ready(Some(cmd)) = this.command_rx.poll_next_unpin(cx) {
+                this.on_command(cmd);
+            }
+        }
+
+        Poll::Pending
+    }
```

Comment:
Fixed busy polling compared with `pool_manager.rs`
Before it would
- Loop 30 times with work counter
- Manually wake task after processing single events
- Only drain one item at a time

Reply:
If we hit into any perf issues here there's probably a better way to poll these events (instead of draining each queue 1 by one). Let's keep it in mind